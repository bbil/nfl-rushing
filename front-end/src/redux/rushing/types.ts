
export interface NflRushingDataItem {
    name: string;
    team: string;
    position: string;

    rushingAttempts: number;
    rushingAttemptsPerGame: number;
    
    totalRushingYards: number;
    rushingYardsPerAttempt: number;
    rushingYardsPerGame: number;

    totalRushingTouchdowns: number;

    longestRush: number;
    longestRushTouchdown: boolean;

    rushFirstDowns: number;
    rushFirstDownPercent: number;

    rush20Plus: number;
    rush40Plus: number;

    rushFumbles: number;
}

export enum SortDirection {
    ASC = 'asc',
    DESC = 'desc'
}

export enum SortOption {
    TOTAL_RUSHING_YARDS = 'total_rushing_yards',
    LONGEST_RUSH = 'longest_rush',
    TOTAL_RUSHING_TOUCHDOWNS = 'total_rushing_touchdowns'
}

export interface ApiQueryParams {
    page: number;
    sort_direction?: SortDirection;
    sort_option?: SortOption;
    name_filter?: string;
}

export interface RushingState {
    data: NflRushingDataItem[];
    pageNumber: number;
    numberOfPages: number;

    sortDirection: SortDirection | null;
    sortOption: SortOption | null;
    nameFilter: string | null;
}
