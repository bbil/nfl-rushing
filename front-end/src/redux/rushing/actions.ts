import { createAction } from 'typesafe-actions';
import { NflRushingDataItem, SortDirection, SortOption } from './types';

export const setData = createAction('rushing/SET_DATA', action => 
    (data: NflRushingDataItem[], pageNumber: number, numberOfPages: number) => action({ data, pageNumber, numberOfPages })
);

export const setSort = createAction('rushing/SET_SORT', action =>
    (sortOption: SortOption, sortDirection: SortDirection) => action({ sortOption, sortDirection })
);

export const setFilter = createAction('rushing/SET_FILTER', action =>
    (filter: string) => action({ filter })
);
