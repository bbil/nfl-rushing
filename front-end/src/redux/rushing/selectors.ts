import { BASE_URL } from "../../constants";
import { State } from "../store";
import { ApiQueryParams, NflRushingDataItem } from "./types";

export const getData = (state: State): NflRushingDataItem[] => state.rushing.data;

export const getApiQueryParams = (state: State): ApiQueryParams => {
    const params: ApiQueryParams = {
        page: state.rushing.pageNumber
    };

    if (state.rushing.sortDirection) {
        params.sort_direction = state.rushing.sortDirection;
    }

    if (state.rushing.sortOption) {
        params.sort_option = state.rushing.sortOption;
    }

    if (state.rushing.nameFilter) {
        params.name_filter = state.rushing.nameFilter;
    }

    return params;
};

export const getHasPreviousPage = (state: State) => state.rushing.pageNumber !== 1;

export const getHasNextPage = (state: State) => state.rushing.pageNumber < state.rushing.numberOfPages;

export const getCsvLink = (state: State): string => {
    const obj: any = {};

    if (state.rushing.sortDirection) {
        obj['sort_direction'] = state.rushing.sortOption;
    }

    if (state.rushing.sortOption) {
        obj['sort_option'] = state.rushing.sortOption;
    }

    if (state.rushing.nameFilter) {
        obj['name_filter'] = state.rushing.nameFilter;
    }

    if (Object.keys(obj).length === 0) {
        return `${BASE_URL}/nfl-rushing/csv`;
    }

    // @ts-ignore
    const queryString = new URLSearchParams(obj).toString();
    return `${BASE_URL}/nfl-rushing/csv?${queryString}`;
}

export const getSorting = (state: State) => {
    return {
        sortDirection: state.rushing.sortDirection,
        sortOption: state.rushing.sortOption
    }
}
