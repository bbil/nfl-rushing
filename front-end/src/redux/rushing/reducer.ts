import { Reducer } from 'redux';
import { ActionType, getType } from 'typesafe-actions';
import { RushingState } from './types';

import * as actions from './actions';

type Actions = ActionType<typeof actions>;

const DEFAULT_STATE: RushingState = {
    data: [],
    pageNumber: 1,
    numberOfPages: 1,
    sortDirection: null,
    sortOption: null,
    nameFilter: null
};

const reducer: Reducer<RushingState, Actions> = (state = DEFAULT_STATE, action) => {
    switch(action.type) {
        case getType(actions.setData):
            return {
                ...state,
                data: action.payload.data,
                numberOfPages: action.payload.numberOfPages,
                pageNumber: action.payload.pageNumber
            }
        case getType(actions.setFilter):
            return {
                ...state,
                nameFilter: action.payload.filter,
                pageNumber: 1
            };
        case getType(actions.removeFilter):
            return {
                ...state,
                nameFilter: null,
                pageNumber: 1
            }
        case getType(actions.setSort):
            return {
                ...state,
                sortDirection: action.payload.sortDirection,
                sortOption: action.payload.sortOption,
                pageNumber: 1
            };
        case getType(actions.removeSort):
            return {
                ...state,
                sortDirection: null,
                sortOption: null,
                pageNumber: 1
            }
        
        default:
            return state;
    }

};

export default reducer;