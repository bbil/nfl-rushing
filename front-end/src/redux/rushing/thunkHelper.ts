import { fetchRushingStats } from "../../api";
import { setData } from "./actions";
import { ApiQueryParams } from "./types";

/**
 * Originally this doFetch helper was defined in thunks.ts; ran into issues with jest spying, which did not
 * like spying on something part of the same module under test. Had a solution that worked for the tests but broke
 * at run-time, so split into it's own file
 */

/**
 * Helper that calls to api, and dispatches action to set data in redux
 * 
 * @param params 
 * @param dispatch 
 */
export async function doFetch(params: ApiQueryParams, dispatch: Function) {
    const rushingStats = await fetchRushingStats(params);
    dispatch(setData(rushingStats.data, rushingStats.pageNumber, rushingStats.numberOfPages));
}