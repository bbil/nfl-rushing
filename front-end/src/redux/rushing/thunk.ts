
import { fetchRushingStats } from "../../api";
import { removeFilter, removeSort, setData, setFilter, setSort } from "./actions";
import { getApiQueryParams } from './selectors';
import { ApiQueryParams, SortDirection, SortOption } from "./types";

/**
 * Helper that calls to api, and dispatches action to set data in redux
 * @param params 
 * @param dispatch 
 */
export async function doFetch(params: ApiQueryParams, dispatch: Function) {
    const rushingStats = await fetchRushingStats(params);
    dispatch(setData(rushingStats.data, rushingStats.pageNumber, rushingStats.numberOfPages));
}

export const initialFetch = () => async (dispatch: Function, getState: any) => {
    const params = getApiQueryParams(getState());

    await exports.doFetch(params, dispatch);
};

export const nextPage = () => async (dispatch: Function, getState: any) => {
    const params = getApiQueryParams(getState());

    params.page += 1;

    await exports.doFetch(params, dispatch);
};

export const previousPage = () => async (dispatch: Function, getState: any) => {
    const params = getApiQueryParams(getState());
    params.page -= 1;

    await exports.doFetch(params, dispatch);
};

export const applyFilter = (nameFilter: string) => async (dispatch: Function, getState: any) => {
    dispatch(setFilter(nameFilter));

    const params = getApiQueryParams(getState());
    await exports.doFetch(params, dispatch);
};

export const clearFilter = () => async (dispatch: Function, getState: any) => {
    dispatch(removeFilter());

    const params = getApiQueryParams(getState());
    await exports.doFetch(params, dispatch);
};

export const applySort = (sortOption: SortOption, sortDirection: SortDirection) => async (dispatch: Function, getState: any) => {
    dispatch(setSort(sortOption, sortDirection));

    const params = getApiQueryParams(getState());
    await exports.doFetch(params, dispatch);
};

export const clearSort = () => async (dispatch: Function, getState: any) => {
    dispatch(removeSort());

    const params = getApiQueryParams(getState());
    await exports.doFetch(params, dispatch);
};