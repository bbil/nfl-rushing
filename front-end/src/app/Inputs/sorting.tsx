import React from 'react';
import { connect } from "react-redux";
import { getSorting } from "../../redux/rushing/selectors";
import { applySort, clearSort } from "../../redux/rushing/thunk";
import { SortDirection, SortOption } from "../../redux/rushing/types";
import { State } from "../../redux/store";

interface SortingDispatchProps {
    applySort: (sortOption: SortOption, sortDirection: SortDirection) => void;
    clearSort: () => void;
}

interface SortingStateProps {
    sortOption: SortOption | null;
    sortDirection: SortDirection | null;
}

export type SortingProps = SortingDispatchProps & SortingStateProps;

export const SortingComponent: React.FC<SortingProps> = props => {
    const [sortOption, setSortOption] = React.useState<SortOption | null>(null)

    const sortOptionOnChange = (event: any) => {
        setSortOption(event.target.value);
    }

    const [sortDirection, setSortDirection] = React.useState<SortDirection | null>(null);

    const sortDirectionOnChange = (event: any) => {
        setSortDirection(event.target.value);
    };

    const clearSort = () => {
        props.clearSort();

        setSortOption(null);
        setSortDirection(null);
    }

    const enableButtons = sortOption !== null && sortDirection !== null;

    return (
        <div>
            <div id='sort-options' onChange={sortOptionOnChange}>
                <input type="radio" value={SortOption.LONGEST_RUSH} name="sortOption" checked={sortOption === SortOption.LONGEST_RUSH} /> Longest Rush
                <input type="radio" value={SortOption.TOTAL_RUSHING_TOUCHDOWNS} name="sortOption" checked={sortOption === SortOption.TOTAL_RUSHING_TOUCHDOWNS} /> Total Rushing Touchdowns
                <input type="radio" value={SortOption.TOTAL_RUSHING_YARDS} name="sortOption" checked={sortOption === SortOption.TOTAL_RUSHING_YARDS}/> Total Rushing Yards
            </div>

            <div id='sort-directions' onChange={sortDirectionOnChange}>
                <input type="radio" value={SortDirection.DESC} name="sortDirection" checked={sortDirection === SortDirection.DESC} /> Desc
                <input type="radio" value={SortDirection.ASC} name="sortDirection" checked={sortDirection === SortDirection.ASC} /> Asc
            </div>

            {/* Ignore null; the enable buttons boolean guarantees it's good */}
            <button id='sort-apply' disabled={!enableButtons} onClick={() => props.applySort(sortOption!, sortDirection!)}>Apply Sort</button>
            <button id='sort-clear' onClick={clearSort}>Clear Sort</button>
        </div>
    )
};

function mapStateToProps(state: State): SortingStateProps {
    const sorting = getSorting(state);
    return {
        sortDirection: sorting.sortDirection,
        sortOption: sorting.sortOption
    }
}

function mapDispatchToProps(dispatch: Function): SortingDispatchProps {
    return {
        applySort: (sortOption: SortOption, sortDirection: SortDirection) => dispatch(applySort(sortOption, sortDirection)),
        clearSort: () => dispatch(clearSort())
    }
}

export const Sorting = connect(mapStateToProps, mapDispatchToProps)(SortingComponent);