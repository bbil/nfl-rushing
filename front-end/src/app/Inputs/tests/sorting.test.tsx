import '@testing-library/react';

import { shallow } from 'enzyme';
import { SortDirection, SortOption } from '../../../redux/rushing/types';

import { SortingComponent, SortingProps } from '../sorting';

describe('Sorting component', () => {
    const PROPS: SortingProps = {
        applySort: jest.fn(),
        clearSort: jest.fn(),
        sortDirection: null,
        sortOption: null
    };

    it('renders default, with nothing checked', () => {
        const wrapper = shallow(<SortingComponent {...PROPS} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('renders with one group checked, button disabled', () => {
        const wrapper = shallow(<SortingComponent {...PROPS} />);

        wrapper.find('#sort-options').simulate('change', { target: { value: SortOption.LONGEST_RUSH }});

        expect(wrapper).toMatchSnapshot();
    });

    it('renders with both grouops checked, button becomes enabled', () => {
        const wrapper = shallow(<SortingComponent {...PROPS} />);

        wrapper.find('#sort-options').simulate('change', { target: { value: SortOption.LONGEST_RUSH }});
        wrapper.find('#sort-directions').simulate('change', { target: { value: SortDirection.ASC }});

        expect(wrapper).toMatchSnapshot();
    });

    it('calls apply sort when button is clicked', () => {
        const wrapper = shallow(<SortingComponent {...PROPS} />);

        wrapper.find('#sort-options').simulate('change', { target: { value: SortOption.LONGEST_RUSH }});
        wrapper.find('#sort-directions').simulate('change', { target: { value: SortDirection.ASC }});

        wrapper.find('#sort-apply').simulate('click');

        expect(PROPS.applySort).toHaveBeenCalledWith(SortOption.LONGEST_RUSH, SortDirection.ASC);
    });

    it('calls clear sort when clear button is clicked, radio buttons no longer checked', () => {
        const wrapper = shallow(<SortingComponent {...PROPS} />);

        wrapper.find('#sort-options').simulate('change', { target: { value: SortOption.LONGEST_RUSH }});
        wrapper.find('#sort-directions').simulate('change', { target: { value: SortDirection.ASC }});

        // double check that radio buttons are checked -- this snapshot should have the radio buttons checked
        expect(wrapper).toMatchSnapshot();

        wrapper.find('#sort-clear').simulate('click');

        expect(PROPS.clearSort).toHaveBeenCalledTimes(1);

        expect(wrapper).toMatchSnapshot();
    });
});
