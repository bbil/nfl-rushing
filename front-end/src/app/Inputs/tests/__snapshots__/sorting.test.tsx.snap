// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorting component calls clear sort when clear button is clicked, radio buttons no longer checked 1`] = `
<div>
  <div
    id="sort-options"
    onChange={[Function]}
  >
    <input
      checked={true}
      name="sortOption"
      type="radio"
      value="longest_rush"
    />
     Longest Rush
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_touchdowns"
    />
     Total Rushing Touchdowns
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_yards"
    />
     Total Rushing Yards
  </div>
  <div
    id="sort-directions"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="desc"
    />
     Desc
    <input
      checked={true}
      name="sortDirection"
      type="radio"
      value="asc"
    />
     Asc
  </div>
  <button
    disabled={false}
    id="sort-apply"
    onClick={[Function]}
  >
    Apply Sort
  </button>
  <button
    id="sort-clear"
    onClick={[Function]}
  >
    Clear Sort
  </button>
</div>
`;

exports[`Sorting component calls clear sort when clear button is clicked, radio buttons no longer checked 2`] = `
<div>
  <div
    id="sort-options"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="longest_rush"
    />
     Longest Rush
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_touchdowns"
    />
     Total Rushing Touchdowns
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_yards"
    />
     Total Rushing Yards
  </div>
  <div
    id="sort-directions"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="desc"
    />
     Desc
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="asc"
    />
     Asc
  </div>
  <button
    disabled={true}
    id="sort-apply"
    onClick={[Function]}
  >
    Apply Sort
  </button>
  <button
    id="sort-clear"
    onClick={[Function]}
  >
    Clear Sort
  </button>
</div>
`;

exports[`Sorting component renders default, with nothing checked 1`] = `
<div>
  <div
    id="sort-options"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="longest_rush"
    />
     Longest Rush
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_touchdowns"
    />
     Total Rushing Touchdowns
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_yards"
    />
     Total Rushing Yards
  </div>
  <div
    id="sort-directions"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="desc"
    />
     Desc
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="asc"
    />
     Asc
  </div>
  <button
    disabled={true}
    id="sort-apply"
    onClick={[Function]}
  >
    Apply Sort
  </button>
  <button
    id="sort-clear"
    onClick={[Function]}
  >
    Clear Sort
  </button>
</div>
`;

exports[`Sorting component renders with both grouops checked, button becomes enabled 1`] = `
<div>
  <div
    id="sort-options"
    onChange={[Function]}
  >
    <input
      checked={true}
      name="sortOption"
      type="radio"
      value="longest_rush"
    />
     Longest Rush
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_touchdowns"
    />
     Total Rushing Touchdowns
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_yards"
    />
     Total Rushing Yards
  </div>
  <div
    id="sort-directions"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="desc"
    />
     Desc
    <input
      checked={true}
      name="sortDirection"
      type="radio"
      value="asc"
    />
     Asc
  </div>
  <button
    disabled={false}
    id="sort-apply"
    onClick={[Function]}
  >
    Apply Sort
  </button>
  <button
    id="sort-clear"
    onClick={[Function]}
  >
    Clear Sort
  </button>
</div>
`;

exports[`Sorting component renders with one group checked, button disabled 1`] = `
<div>
  <div
    id="sort-options"
    onChange={[Function]}
  >
    <input
      checked={true}
      name="sortOption"
      type="radio"
      value="longest_rush"
    />
     Longest Rush
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_touchdowns"
    />
     Total Rushing Touchdowns
    <input
      checked={false}
      name="sortOption"
      type="radio"
      value="total_rushing_yards"
    />
     Total Rushing Yards
  </div>
  <div
    id="sort-directions"
    onChange={[Function]}
  >
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="desc"
    />
     Desc
    <input
      checked={false}
      name="sortDirection"
      type="radio"
      value="asc"
    />
     Asc
  </div>
  <button
    disabled={true}
    id="sort-apply"
    onClick={[Function]}
  >
    Apply Sort
  </button>
  <button
    id="sort-clear"
    onClick={[Function]}
  >
    Clear Sort
  </button>
</div>
`;
